Add missing controls to flex layouts.

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Fixes #45868 and #46133 as well as making #45117 possible. Complementary to #47099 - technically this PR provides the controls to solve all the layout problems in those issues, which #47099 addresses the problem of which is the most sensible default aligmnent value for flex containers.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Adds a "stretch" option to the vertical alignment controls for the Row block:
<img width="380" alt="Screenshot 2023-01-13 at 3 36 19 pm" src="https://user-images.githubusercontent.com/8096000/212238569-792923a1-d115-4a7f-830b-5d3ef46e551e.png">

* Adds vertical alignment controls to the Stack block, with "top", "center", "bottom" and "space-between" options (these essentially implement `justify-content` for Stack):
<img width="270" alt="Screenshot 2023-01-13 at 3 36 41 pm" src="https://user-images.githubusercontent.com/8096000/212238521-d5609e69-2116-4a50-80bd-a83ba1d401fd.png">

* Adds a "stretch" option to the justification controls for the Stack block:
<img width="283" alt="Screenshot 2023-01-13 at 3 35 34 pm" src="https://user-images.githubusercontent.com/8096000/212238537-c479f44c-4921-49ed-bad9-71b6ef9f7e9e.png">

## Todo:

We need some icons for "stretch" and also for "space-between" in a vertical context, cc @WordPress/gutenberg-design 

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Add Row and Stack blocks with content to a post;
2. Verify that it's possible to justify and align contents in both orientations equally.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

tellthemachines committed:
Add missing controls to flex layouts.

andrewserong commented on packages/block-editor/src/layouts/flex.js:
```diff
@@ -44,12 +44,15 @@ const alignItemsMap = {
 	left: 'flex-start',
 	right: 'flex-end',
 	center: 'center',
+	stretch: 'stretch',
```
Do we also need to update the corresponding allowed values in `layout.php` around here: https://github.com/WordPress/gutenberg/blob/a41ea8177b7ac9c2b1861bbacf5768ec459195b0/lib/block-supports/layout.php#L190 ?

tellthemachines replied:
I just pushed the update!

tellthemachines committed:
Front end support

andrewserong commented on packages/block-editor/src/layouts/flex.js:
```diff
@@ -44,12 +44,15 @@ const alignItemsMap = {
 	center: 'center',
 	bottom: 'flex-end',
+	stretch: 'stretch',
+	spaceBetween: 'space-between',
```
Totally not urgent, but the horizontal alignment uses the kebab case to match the CSS value, rather than the camelCase JS version of space between. Should we make these consistent and use `space-between` here, too, (and in the PHP code) for the value we store in block attributes?

tellthemachines replied:
Done!

jameskoster commented:
> Adds a "stretch" option to the vertical alignment controls for the Row block:

This one works well and seems to be an implementation of `align-items`.

> Adds vertical alignment controls to the Stack block, with "top", "center", "bottom" and "space-between" options (these essentially implement justify-content for Stack):

This one works less well, because the settings only have any effect when the containing Stack is taller that the sum of its parts. Since it's not possible to define a custom height on a Stack, that's a rare occurrence unless I'm missing something? Most of the time you'll see this, which feels kind of broken:

![align](https://user-images.githubusercontent.com/846565/212340355-135a3535-6aba-4ac6-9949-37144d749c34.gif)

Edit: I suppose if a Stack is placed inside a Row which has stretch-to-fill applied, then this setting would then have an effect. So I guess the question is whether we should attempt to detect that condition before displaying the control, or not. 

> Adds a "stretch" option to the justification controls for the Stack block:

This one works well too, with the except of blocks that have a specific width applied, like Images. For example this seems a bit strange:

<img width="1368" alt="Screenshot 2023-01-13 at 14 17 49" src="https://user-images.githubusercontent.com/846565/212341045-c4e6e703-da52-4c34-8d4c-56948922fdd5.png">

The block is clearly stretching as intended, but the image retains its width. I'm not entirely sure what to do here, but I think my expectation would be that `stretch` overrides `width`. 

---

The icons need some thought, and potentially a separate discussion.

jameskoster commented:
Discussion for icons here: https://github.com/WordPress/gutenberg/discussions/47153

jasmussen commented:
> Discussion for icons here: https://github.com/WordPress/gutenberg/discussions/47153

Outside of redesigning all the icons (which can still happen, but could potentially happen on a separate track), am I correct that the primary icons missing are a new vertical space-between, and a new vertical stretch to fill?

For vertical space between can we just rotate the horizontal space-between, and then also make a version of that which stretches the square inside to fill for the other? Might not be perfect, but might unblock this one, buying time to explore the new icons better? (I.e. it might be a few mins. of work to make those two.)

jameskoster commented:
If we're happy with the `space-between` and `stretch` suggestions in https://github.com/WordPress/gutenberg/discussions/47153, perhaps we can add those here, then update the existing icons separately?

jameskoster commented:
From the discussion in https://github.com/WordPress/gutenberg/discussions/47153, here are the svgs:

**Space-between horizontal**

```
<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M12.75 4H11.25V20H12.75V4ZM4 7L8 7V17L4 17V7ZM16 7L20 7V17L16 17V7Z"/>
</svg>
```

**Space-between vertical**

```
<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M7 4H17V8L7 8V4ZM7 16L17 16V20L7 20V16ZM20 11.25H4V12.75H20V11.25Z" />
</svg>
```

**Stretch horizontal**
```
<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M4 4H5.5V20H4V4ZM7 10L17 10V14L7 14V10ZM20 4H18.5V20H20V4Z"/>
</svg>
```

**Stretch vertical**

```
<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M4 4L20 4L20 5.5L4 5.5L4 4ZM10 7L14 7L14 17L10 17L10 7ZM20 18.5L4 18.5L4 20L20 20L20 18.5Z"/>
</svg>
```

ndiego commented:
> This one works less well, because the settings only have any effect when the containing Stack is taller that the sum of its parts. Since it's not possible to define a custom height on a Stack, that's a rare occurrence unless I'm missing something?

@jameskoster you can set min-height on Stacks now.

andrewserong commented:
Just noticed that it looks like the PHP unit tests for `layout.php` need to be updated ([this test data](https://github.com/WordPress/gutenberg/blob/9ee17a79210e9010fb774c841de559c2cf26d2fe/phpunit/block-supports/layout-test.php#L365)) now that the output includes the extra `justify-content:flex-end;` for the vertical orientation.

Also, in case an earlier comment got lost in the backscroll, what do you think about updating the value for space between so that it's kebab-case when serialized to the block markup to match the existing `space-between` value used for horizontal layouts?

tellthemachines committed:
Change space-between to kebab case and fix popover props warning.

tellthemachines committed:
Update to use the correct icons.

tellthemachines committed:
Update PHP test output.

tellthemachines commented on packages/block-editor/src/components/block-vertical-alignment-control/ui.js:
```diff
@@ -49,7 +59,7 @@ function BlockVerticalAlignmentUI( {
 	const extraProps = isToolbar
 		? { isCollapsed }
-		: { popoverProps: { POPOVER_PROPS } };
+		: { popoverProps: { variant: 'toolbar' } };
```
This change fixes the following React warning in the console:
```
Warning: React does not recognize the `POPOVER_PROPS` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `popover_props` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
```

tellthemachines committed:
Remove todo comment.

tellthemachines commented:
Thanks for the feedback folks! And for providing the new icons too 😄 

I think I've addressed all the remaining feedback.

> The block is clearly stretching as intended, but the image retains its width. I'm not entirely sure what to do here, but I think my expectation would be that stretch overrides width.

Yeah, this is a tough one because the width is set inline on the image block, and also because we may not want to auto-enlarge an image with a small intrinsic size, as it'll end up looking fuzzy or pixelated, which might not be intended 😅 

There's a similar issue with the Button blocks, because they have their own width settings. We should probably look into what the desired behaviour is there, too! But I think these would best be addressed as follow-ups as they'll likely be a bit of back and forth until we reach the optimal solution.

ndiego commented:
@tellthemachines are we aiming for 6.2 with PR?

MaggieCabrera commented:
Not sure if this is related to this PR or not, but while testing (which btw it was working very nicely for me), I saw that the column block doesn't have the alignment controls on the settings tab, just the toolbar:

<img width="1702" alt="Screenshot 2023-01-27 at 13 13 35" src="https://user-images.githubusercontent.com/3593343/215084691-14a4366e-8b55-4162-ae19-46fb5ec814a9.png">

ndiego reviewed (approved):
I'm approving from a functionality perspective. It's working really well in my testing. Great work 🙌 

![image](https://user-images.githubusercontent.com/4832319/215125583-f61d9179-0a43-4148-ac94-7a25d506b029.png)

This is likely a separate PR, but it would be nice to have the vertical alignment in the settings tab as well. Much like @MaggieCabrera mentions above regarding the Columns block.

![image](https://user-images.githubusercontent.com/4832319/215126006-c9349db1-d9fe-4a8e-ad29-d2c6261c3e5a.png)

bgardner commented:
A big +1 for this PR, as I am working on a design that needs this functionality. I hope it makes it into 15.1 and 6.2.

andrewserong commented on packages/block-editor/src/layouts/flex.js:
```diff
@@ -293,6 +313,13 @@ function FlexLayoutJustifyContentControl( {
 			label: __( 'Space between items' ),
 		} );
+	} else {
+		// Todo: we also need an icon here.
```
Can this comment be removed now?

tellthemachines commented:
> @tellthemachines are we aiming for 6.2 with PR?

Yes! 

> Not sure if this is related to this PR or not, but while testing (which btw it was working very nicely for me), I saw that the column block doesn't have the alignment controls on the settings tab, just the toolbar:

The alignment controls don't appear in the sidebar for any blocks yet. That's definitely something for another PR, as it'll need some design guidance 😄 

Thanks for reviewing and testing, folks!

andrewserong reviewed (commented):
This is mostly working really well for me @tellthemachines! I found a few little issues:

When toggling between horizontal and vertical, with a content justification value not present in the other orientation, the icon looks partially selected even though it kind of isn't. What should happen when we switch orientations here?

![2023-01-30 10 34 15](https://user-images.githubusercontent.com/14988353/215362561-ffcc3d88-2029-47e9-a151-afb4b8080a7d.gif)

I noticed with the Buttons block, that the stretch item doesn't show up in the toolbar justification dropdown, so the selected icon appears to be the justified icon (the icon in blue in the toolbar):

![image](https://user-images.githubusercontent.com/14988353/215362681-d6fb623a-fabd-4136-afe0-d24cfa7baaf2.png)

Another issue with the Button block's styling is that it doesn't _quite_ appear to handle the stretch option, since the Button is styled on the child `a` tag. So, the button block itself `.wp-block-button` is stretched to be full-width, however its child isn't.

![image](https://user-images.githubusercontent.com/14988353/215362955-ed20e387-f22e-421d-b37b-31ca904e58b1.png)


Would it be worth adding a rule to the Button block's CSS to handle the stretch case? I imagine it could be a similar rule to the existing rule for `.has-custom-width` ([here](https://github.com/WordPress/gutenberg/blob/4beedebc396403fa5a2f6ac30681f5f13a5db734/packages/block-library/src/button/style.scss#L43)), where we'd set `.wp-block-button__link` to `width:100%` if `.wp-block-buttons.is-vertical.is-content-justification-stretch`? Though there could be a simpler rule for it 🤔

These probably aren't blockers to merging though, and could be looked at in follow-ups, I think.

tellthemachines commented:
> When toggling between horizontal and vertical, with a content justification value not present in the other orientation, the icon looks partially selected even though it kind of isn't. What should happen when we switch orientations here?

I'm not sure! Maybe it would be best to switch to default (flex-start).

I'll do a quick follow-up to address the remaining issues.

andrewserong commented:
> I'm not sure! Maybe it would be best to switch to default (flex-start).

That sounds good to me 👍

Thanks for following up!
